core/dns.go

```diff
@@ -32,23 +34,35 @@ var (
 		IP:   []byte{192, 168, 0, 0},
 		Mask: []byte{255, 255, 0, 0},
 	}
+
+	// RFC5735
+	// 127.0.0.0/8
+	rfc5735_127 = net.IPNet{
+		IP:   []byte{127, 0, 0, 0},
+		Mask: []byte{255, 0, 0, 0},
+	}
 )
 
 // DNSResolverImpl represents a client that talks to an external resolver
 type DNSResolverImpl struct {
-	DNSClient *dns.Client
-	Servers   []string
+	DNSClient              *dns.Client
+	Servers                []string
+	allowLoopbackAddresses bool
 }
 
 // NewDNSResolverImpl constructs a new DNS resolver object that utilizes the
 // provided list of DNS servers for resolution.
-func NewDNSResolverImpl(dialTimeout time.Duration, servers []string) *DNSResolverImpl {
+func NewDNSResolverImpl(dialTimeout time.Duration, servers []string, allowLoopback bool) *DNSResolverImpl {
```

Comment:
So this blind boolean is a little ugh. It's hard to tell what it means when all you've got is a call to NewDNSResolverImpl(someDur, servers, true)

It would be cool for this code to either get a NewDNSResolverImplForTest() that did a call to a private constructor function with this boolean to set to true or for a list of banned IP addresses to be passed in here.

For the former idea, you'd still have to do an if statement in the binaries but that's fine.